### ISSUE TITLE
Test scores remain exactly the same when loss function for Physnet changed

### ISSUE THREAD

**Issue by deepshiksharma** [ubicomplab/rPPG-Toolbox#377]
I tried changing the default loss function for Physnet, and every time I get exactly the same test scores. I even tried MSE and MAE loss and it remains exactly the same. Please help : (

This is how I have set up the configuration file:
```
BASE: ['']
TOOLBOX_MODE: "train_and_test" # "train_and_test"  or "only_test"
TRAIN:
  BATCH_SIZE: 4
  EPOCHS: 30
  LR: 9e-3
  MODEL_FILE_NAME: UBFC_UBFC_PURE_physnet_stan
  PLOT_LOSSES_AND_LR: True
  DATA:
    FS: 30
    DATASET: UBFC-rPPG
    DO_PREPROCESS: False            # if first time, should be true
    DATA_FORMAT: NCDHW
    DATA_PATH: "./../rPPG-Toolbox/RawData/"                     # Raw dataset path, need to be updated
    CACHED_PATH: "PreprocessedData/"    # Processed dataset save path, need to be updated
    FILE_LIST_PATH: "DataFileLists/"    # Path to store file lists, needs to be updated
    EXP_DATA_NAME: ""
    BEGIN: 0.0
    END: 0.8
    PREPROCESS:
      DATA_TYPE: ['DiffNormalized'] #if use physnet, should be DiffNormalized
      DATA_AUG: ['None']    # 'None' or 'Motion' is supported, used if the data path points to an augmented dataset or requires augmentation
      LABEL_TYPE: DiffNormalized
      DO_CHUNK: True
      CHUNK_LENGTH: 128  #only support for factor of 512
      CROP_FACE:
        DO_CROP_FACE: True
        BACKEND: 'HC'    # HC for Haar Cascade, RF for RetinaFace
        USE_LARGE_FACE_BOX: True
        LARGE_BOX_COEF: 1.5
        DETECTION:
          DO_DYNAMIC_DETECTION: False
          DYNAMIC_DETECTION_FREQUENCY : 32
          USE_MEDIAN_FACE_BOX: False    # This should be used ONLY if dynamic detection is used
      RESIZE:
        H: 72
        W: 72
VALID:
  DATA:
    FS: 30
    DATASET: UBFC-rPPG
    DO_PREPROCESS: False                # if first time, should be true
    DATA_FORMAT: NCDHW
    DATA_PATH: "./../rPPG-Toolbox/RawData/"                     # Raw dataset path, need to be updated
    CACHED_PATH: "PreprocessedData/"    # Processed dataset save path, need to be updated
    FILE_LIST_PATH: "DataFileLists/"    # Path to store file lists, needs to be updated
    EXP_DATA_NAME: ""
    BEGIN: 0.8
    END: 0.9
    PREPROCESS:
      DATA_TYPE: ['DiffNormalized'] #if use physnet, should be DiffNormalized
      DATA_AUG: ['None']    # 'None' or 'Motion' is supported, used if the data path points to an augmented dataset or requires augmentation
      LABEL_TYPE: DiffNormalized
      DO_CHUNK: True
      CHUNK_LENGTH: 128  #only support for factor of 512
      CROP_FACE:
        DO_CROP_FACE: True
        BACKEND: 'HC'    # HC for Haar Cascade, RF for RetinaFace
        USE_LARGE_FACE_BOX: True
        LARGE_BOX_COEF: 1.5
        DETECTION:
          DO_DYNAMIC_DETECTION: False
          DYNAMIC_DETECTION_FREQUENCY : 32
          USE_MEDIAN_FACE_BOX: False    # This should be used ONLY if dynamic detection is used
      RESIZE:
        H: 72
        W: 72
TEST:
  OUTPUT_SAVE_DIR: "Test_Outputs"
  METRICS: ['MAE', 'RMSE', 'MAPE', 'Pearson', 'SNR', 'BA']
  USE_LAST_EPOCH: True                   # to use provided validation dataset to find the best epoch, should be false
  DATA:
    FILTERING:
      USE_EXCLUSION_LIST: True
      EXCLUSION_LIST: [
      's3_T1', 's8_T1', 's9_T1', 's26_T1', 's28_T1', 's30_T1', 's31_T1', 's32_T1',
      's33_T1', 's40_T1', 's52_T1', 's53_T1', 's54_T1', 's56_T1', 's1_T2', 's4_T2',
      's6_T2', 's8_T2', 's9_T2', 's11_T2', 's12_T2', 's13_T2', 's14_T2', 's19_T2',
      's21_T2', 's22_T2', 's25_T2', 's26_T2', 's27_T2', 's28_T2', 's31_T2', 's32_T2',
      's33_T2', 's35_T2', 's38_T2', 's39_T2', 's41_T2', 's42_T2', 's45_T2', 's47_T2',
      's48_T2', 's52_T2', 's53_T2', 's55_T2', 's5_T3', 's8_T3', 's9_T3', 's10_T3',
      's13_T3', 's14_T3', 's17_T3', 's22_T3', 's25_T3', 's26_T3', 's28_T3', 's30_T3',
      's32_T3', 's33_T3', 's35_T3', 's37_T3', 's40_T3', 's47_T3', 's48_T3', 's49_T3',
      's50_T3', 's52_T3', 's53_T3']
      SELECT_TASKS: True
      TASK_LIST: ['T1', 'T2', 'T3']
    FS: 30
    DATASET: UBFC-rPPG
    DO_PREPROCESS: False                  # if first time, should be true
    DATA_FORMAT: NCDHW
    DATA_PATH: "./../rPPG-Toolbox/RawData/"                     # Raw dataset path, need to be updated
    CACHED_PATH: "PreprocessedData/"    # Processed dataset save path, need to be updated
    FILE_LIST_PATH: "DataFileLists/"    # Path to store file lists, needs to be updated
    EXP_DATA_NAME: ""
    BEGIN: 0.9
    END: 1.0
    PREPROCESS:
      DATA_TYPE: ['DiffNormalized'] #if use physnet, should be DiffNormalized
      LABEL_TYPE: DiffNormalized
      DO_CHUNK: True
      CHUNK_LENGTH: 128  #only support for factor of 512
      CROP_FACE:
        DO_CROP_FACE: True
        BACKEND: 'HC'    # HC for Haar Cascade, RF for RetinaFace
        USE_LARGE_FACE_BOX: True
        LARGE_BOX_COEF: 1.5
        DETECTION:
          DO_DYNAMIC_DETECTION: False
          DYNAMIC_DETECTION_FREQUENCY : 32
          USE_MEDIAN_FACE_BOX: False    # This should be used ONLY if dynamic detection is used
      RESIZE:
        H: 72
        W: 72
DEVICE: cuda:0
NUM_OF_GPU_TRAIN: 1
LOG:
  PATH: Runs/exp
MODEL:
  DROP_RATE: 0.2
  NAME: Physnet
  PHYSNET:
    FRAME_NUM: 128
INFERENCE:
  BATCH_SIZE: 4
  EVALUATION_METHOD: "FFT"        # "FFT" or "peak detection"
  EVALUATION_WINDOW:
    USE_SMALLER_WINDOW: False        # Change this if you'd like an evaluation window smaller than the test video length
    WINDOW_SIZE: 10        # In seconds
  MODEL_PATH: ""
```

**Comment by yahskapar** [comment]
Can you share your PhysNet trainer you’re using and the changes you’ve made to modify the loss functions there? 

Also, your config file looks a little strange considering it has the exclusion list for UBFC-PHYS, but it looks like you’re doing intra-dataset or within dataset experiments. Maybe try eliminating your config as a potential source of issues and just try using one of the default configs (with your data paths) while changing the loss function?

**Comment by deepshiksharma** [comment]
please help : ((

**Comment by yahskapar** [comment]
Can you inspect the loss plots (generated by keeping `PLOT_LOSSES_AND_LR: True` in your config file) and see if they differ? For example, generate one for using the default `Neg_Pearson` loss, save it, and then generate another for using `MSE`, save it, and compare the two visually (feel free to share the two plots here as well).

From the code you shared, nothing looks glaringly off in terms of how you're setting the loss model. I'm assuming for the time being you commented out the `Neg_pearson` import intentionally, which includes the custom one you are trying to use. It's possible that, despite different loss curves, the impact of the scheduler (`OneCyclerLR`) in particular is significant here and causes roughly the same metric results. In that case, verifying that the loss plots themselves are differing would be a good sanity check, and potential next steps would be trying a different scheduler or perhaps different hyper-parameters (e.g., learning rate) depending on your goal.

**Comment by deepshiksharma** [comment]
I tried a lot of things, and then the scores changed. I was then able to work from there, because now it is not frozen at the exact same values. sorry i panicked. thank you.
With best regards,
An Intern
